src/main/scala/au/com/agiledigital/toolform/command/generate/kubernetes/VolumeClaimWriter.scala

```diff
@@ -0,0 +1,106 @@
+package au.com.agiledigital.toolform.command.generate.kubernetes
+
+import au.com.agiledigital.toolform.model.{Resource}
+import collection.JavaConversions._
+import collection.JavaConverters._
+import com.typesafe.config._
+import cats.implicits._
+import scala.util.{Failure, Success, Try}
+
+object VolumeClaimWriter extends KubernetesWriter {
+
+  // write the various access modes
+  def writeAccessModes(resource: Resource, serviceName: String): Result[Unit] = {
+    val acceptableModes = List("ReadWriteOnce", "ReadOnlyMany")
+    val accessModes = resource.settings match {
+      case Some(aModes) => {
+        Try(aModes.getStringList("accessModes")) match {
+          case Success(list) => list
+          // no accessModes specified in project.conf
+          case Failure(error) => {
+            println(s"Warning: no access modes specified for resource: $serviceName. Default of ReadWriteOnce used")
+            val res: java.util.List[String] = List("ReadWriteOnce")
+            res
+          }
+        }
+      }
+      case None => {
+        println(s"Warning: no access modes specified for resource: $serviceName. Default of ReadWriteOnce used")
+        val res: java.util.List[String] = List("ReadWriteOnce")
+        res
+      }
+    }
+    val modes = String.join("-", accessModes).split("-").toList
+    for (mode <- modes) {
+      if (!acceptableModes.contains(mode)) {
+        println(s"Warning: $mode access mode unsupported for resource: $serviceName")
+      }
+    }
+
+    for {
+      _ <- modes.traverse_(mode => write(s"- $mode"))
+    } yield ()
+  }
+
+  // return the storage size of the disk resource, printing a warning and using default if no size specified
+  def getStorageSize(resource: Resource, serviceName: String): String =
+    resource.storage match {
+      case Some(strg) =>
+        strg match {
+          case "small"  => "2Gi"
+          case "medium" => "5Gi"
+          case "large"  => "10Gi"
+          case _        => strg
+        }
+      case None => { // use default and issue warning
+        println(s"Warning: No disk volume size given for resource: $serviceName. Default size of 2Gi used.")
+        "2Gi"
+      }
+    }
+
+  /**
+    * Writes a Kubernetes "PersistentVolumeClaim" specification based on the provided disk resource.
+    *
+    * A PersistentVolumeClaim is a request for storage by a user
+    *
+    * @see https://kubernetes.io/docs/concepts/storage/persistent-volumes/
+    *
+    * @param resource   the disk resource object that will be used to create the PersistentVolumeClaim.
+    * @return           a state monad encapsulating the context of the writing process after the method has completed.
+    */
+  def writeVolumeClaim(resource: Resource): Result[Unit] = {
```

Comment:
Is there a reason that a volume claim didn't get a first class object here? Could a volume claim be deserialized to its own case class?

E.g. In `ServiceWriter` for example, `writeService` takes a `ToolFormService` object which contains all the information needed to write the service.

Reply:
The reason it is written this way is because VolumeClaims are currently just from a Resources list. Is it worth looking at changing the data model to split Volumes out from Resources?

Reply:
I think that it would make the code cleaner to do this. Currently, a `Resource` object can represent two things: a docker resource OR storage. Because this object is loosely defined, validation leaks into all the writer methods. Previously, these writer methods did have to worry about error handling.

You could extract the common fields between a docker resource and a storage resource into a base class and then make two sub classes `StorageResource` and `DockerResource`. This would make things a bit better because then the validation would be done at parse time and you wouldn't have to make things like `image` optional because a `DockerResource` always should have an `image` but a `StorageResource` doesn't even have that field.
I don't think this is ideal because the only things in common between a storage and a docker resource are `id`, `resourceType` and `settings` so they don't really have anything in common semantically. Also I'm not sure how you would get the PureConfig parser to deserialise to the right class based on `resourceType`.

Therefore I'd probably suggest breaking volumes out into first class objects but I'm not sure what semantics were discussed beforehand (e.g. is a volume semantically a resource?)

Reply:
I had a discussion with @timechanter on Slack and volumes are technically a resource so they probably shoudn't be broken into a separate concept.